### PR TITLE
fix: type coercion for expr/subquery in InSubquery

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #.

# Rationale for this change

# What changes are included in this PR?

 type coercion should do for expr and subquery in InSubquery

# Are these changes tested?

Add UT test for expr/subquery in Insubquery

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/expr/src/expr_schema.rs [hunk 1]
```diff
@@ -326,7 +317,12 @@ impl ExprSchemable for Expr {
     }
 }
 
-fn cast_subquery(subquery: Subquery, cast_to_type: &DataType) -> Result<Subquery> {
+/// cast subquery in InSubquery/ScalarSubquery to a given type.
```
Comment (liukun4515) [inline datafusion/expr/src/expr_schema.rs:320]: I find this method is used in `scalar subquer`, but the comment is for Insubquery/scalarquery.

File: datafusion/optimizer/src/analyzer/type_coercion.rs [hunk 2]
```diff
@@ -147,20 +148,24 @@ impl TreeNodeRewriter for TypeCoercionRewriter {
...
             } => {
-                let expr_type = expr.get_type(&self.schema)?;
                 let new_plan = analyze_internal(&self.schema, &subquery.subquery)?;
+                let expr_type = expr.get_type(&self.schema)?;
                 let subquery_type = new_plan.schema().field(0).data_type();
-                let expr = if &expr_type == subquery_type {
-                    expr
-                } else {
-                    Box::new(expr.cast_to(subquery_type, &self.schema)?)
+                let result_type = comparison_coercion(&expr_type, subquery_type).ok_or(DataFusionError::Plan(
```
Comment (liukun4515) [inline datafusion/optimizer/src/analyzer/type_coercion.rs:154]: maybe the common_type or coercion_type is better.
the result type of `in subquery` or `comparation` is `boolean`.

File: datafusion/optimizer/src/analyzer/type_coercion.rs [hunk 3]
```diff
@@ -147,20 +148,24 @@ impl TreeNodeRewriter for TypeCoercionRewriter {
...
+                    format!(
+                        "expr type {expr_type:?} can't cast to {subquery_type:?} in InSubquery"
+                    ),
+                ))?;
+                let new_subquery = Subquery {
+                    subquery: Arc::new(analyze_internal(
+                        &self.schema,
+                        &subquery.subquery,
```
Comment (liukun4515) [inline datafusion/optimizer/src/analyzer/type_coercion.rs:162]: can we use the `new_plan` which is generated by `analyze_internal`?

File: datafusion/optimizer/src/analyzer/type_coercion.rs [hunk 4]
```diff
@@ -749,16 +754,30 @@ mod test {
     };
     use crate::test::assert_analyzed_plan_eq;
 
+    fn empty() -> Arc<LogicalPlan> {
```
Comment (liukun4515) [inline datafusion/optimizer/src/analyzer/type_coercion.rs:757]: 👍

File: datafusion/optimizer/src/analyzer/type_coercion.rs [hunk 5]
```diff
@@ -1428,33 +1404,47 @@ mod test {
...
+        \n  Subquery:\
+        \n    Projection: CAST(a AS Int64)\
+        \n      EmptyRelation\
+        \n  EmptyRelation";
+        assert_analyzed_plan_eq(Arc::new(TypeCoercion::new()), &plan, expected)?;
+        Ok(())
+    }
+
+    #[test]
+    fn in_subquery_cast_expr() -> Result<()> {
```
Comment (liukun4515) [inline datafusion/optimizer/src/analyzer/type_coercion.rs:1432]: could you please add or replace the cast case for decimal data type?
the coercion type or the common type is not same with `expr` and `subquery`, this case will both cast `expr` and `subquer` together.

**Comment by jackwener** [comment]
thanks @liukun4515 , has resolved them

**Review by liukun4515 — APPROVED** [review]

**Pull request merged**